File: Kotlin/src/main/kotlin/leetcode/easy/383 Ransom Note.kt
package leetcode.easy
/*Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.



Example 1:

Input: ransomNote = "a", magazine = "b"
Output: false
Example 2:

Input: ransomNote = "aa", magazine = "ab"
Output: false
Example 3:

Input: ransomNote = "aa", magazine = "aab"
Output: true


Constraints:

1 <= ransomNote.length, magazine.length <= 105
ransomNote and magazine consist of lowercase English letters.*/
import java.util.*


class `383 Ransom Note` {

    fun canConstruct(ransomNote: String, magazine: String): Boolean {
        val count = magazine.groupingBy { it }.eachCount().toMutableMap()
        for (c in ransomNote) {
            if (c !in count || count[c] == 0) {
                return false
            } else {
                count[c] = count[c]!! - 1
            }
        }
        return true
    }
}

File: Python/AlgoBootcamp/leetcode/easy/383. Ransom Note.py
# Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.
#
# Each letter in magazine can only be used once in ransomNote.
#
#
#
# Example 1:
#
# Input: ransomNote = "a", magazine = "b"
# Output: false
# Example 2:
#
# Input: ransomNote = "aa", magazine = "ab"
# Output: false
# Example 3:
#
# Input: ransomNote = "aa", magazine = "aab"
# Output: true
#
#
# Constraints:
#
# 1 <= ransomNote.length, magazine.length <= 105
# ransomNote and magazine consist of lowercase English letters.
from collections import Counter


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        count = Counter(magazine)
        for c in ransomNote:
            if c not in count or count[c] == 0:
                return False
            else:
                count[c] -= 1
        return True


sol = Solution()
assert sol.canConstruct("aa", "aab") is True
assert sol.canConstruct("aaa", "aab") is False